keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+Currently kubeadm is released as part of the following k8s artifacts:
+- Node tarballs (Linux / Windows)
+- deb/rpm packages
+
+The kubeadm binary is built using either bazel (amd64 only) or using Docker
+(cross-build, all supported architectures).
+
+The binary is built with the same version as the rest of the released binaries
+like the kubelet. The version tooling resides in k/k under the /hack directory
+and it must be copied to k/kubeadm for consistency. The tooling that perform
+the packaging and building are location in k/release and require manual interaction.
+
+##### Building kubeadm from tarballs
+
+For the initial proposal this will not be supported, but if there
+are plans to distribute the kubeadm source code as part of a k8s release,
+the solution would be to write a static KUBE_VERSION file in the tarball.
+
+If GitHub users download the source code directly from the kubeadm repository
+the build process must be able to detect that this is not a Git repository
+and ask the users to provide a KUBE_VERSION file.
+
+##### Bazel vs go build
+
+Bazel was investigated as an option for building kubeadm from the new
+repository but the benefits are unclear. BUILD files present a maintenance
+challenge and also the bazel WORKSPACE and gazelle files need to be kept
+in sync with k8s. On the other hand building using “go build” can take
+around ~20 sec on a fast machine with fast internet.
+
+The majority voted to not use bazel in the k/kubeadm repository.
+
+If caching is required a build container can be used.
+
+##### Syncing tags and branches between k/k and k/kubeadm
+
+A brief discussion of the #prow maintainers revealed the proposal that
+a post-submit job can be used for that instead of prow plugins.
+
+Supposedly such a job will trigger each time a new branch is created
+or if a tag is pushed. This means that the post-submit can trigger any bash
+script or application to get the latest list of branches and tags in k/k
+and apply the same to the kubeadm repository.
+
+Given kubeadm will potentially move within a single cycle this means that
+not all tags and branches have to be synced - see the options in “Versions
+in the support skew”. So potentially there must be a minimum version
+check in the post-submit job.
+
+Scenarios:
+
+###### k/k creates a new release-x.yy branch
+
+The post-submit job checks out master in k/kubeadm, creates a new branch
+release-x.yy and pushes the branch.
+
+###### k/k ‘s release-x.yy branch is fast-forwarded to master
+
+Currently this is a manual step performed by someone on the release team.
+It might be something that the k/kubeadm maintainers have to do manually
+on a daily basis. Automation is possible but is tricky and needs further
+discussion. Can be done as a nice-to-have in the future.
+
+###### k/release starts building kubeadm version x.yy.
+
+There is a chicken and egg problem where a tag in the k/kubeadm repository
+cannot be added before the k8s release is cut. This means that the build
```

Comment:
I'm missing the justification of this statement.   Can you clarify the problem?

Reply:
when a k8s release is cut, a tag is added in k/k, a post-submit job triggers only after a tag is added for k/k. this means a k/kubeadm tag can only happen after k/k is already tagged.

but the k8s release must already include the k/kubeadm artifact, which means that the k/kubeadm artifact must be built from a the latest k/kubeadm commit.

Reply:
Say instead a human adds a tag in k/kubeadm.  Automation builds kubeadm.  K/k is tagged, k8s is released including the tagged k/k content and tagged k/kubeadm and build kubeadm artifact.

We're describing a distribution.  How does the distributor decide which component versions to consume?  It's an unusual model from my perspective to have the distributor start their build and that trigger on demand creation of the components to be consumed, versus the distributor start their build and have that consume all ready components.

Reply:
k/kubeadm versioning does not make sense without the k/k versioning.
therefore it seems reasonable for me to automate this part.

the publishing bot is already doing that for a number of repositories.

Reply:
Discussing with @neolit123 the kubeadm branches will be maintained as release ready and at any point when k/k is to be tagged/built/released, kubeadm can also be similarly tagged/built/released.  We'll see how that goes and evolve as needed.